src/devinterp/slt/gradient.py

```diff
@@ -0,0 +1,151 @@
+        max_diff = abs(new_max - self.max_grad)
+        num_new_max_bins = math.ceil(max_diff / self.bin_size)
+        if num_new_min_bins:
+            for param_name in self.grad_dists:
+                zeros = torch.zeros((self.num_chains, self.num_draws, num_new_min_bins), dtype=torch.int64).to(self.device)
+                self.grad_dists[param_name] = torch.cat([zeros, self.grad_dists[param_name]], dim=2)
+        if num_new_max_bins:
+            for param_name in self.grad_dists:
+                zeros = torch.zeros((self.num_chains, self.num_draws, num_new_max_bins), dtype=torch.int64).to(self.device)
+                self.grad_dists[param_name] = torch.cat([self.grad_dists[param_name], zeros], dim=2)
+        self.min_grad -= num_new_min_bins * self.bin_size
+        self.num_bins += num_new_min_bins + num_new_max_bins
+    
+    def _merge_bins(self):
+        while self.num_bins > 2 * self.min_bins:
+            new_bin_count = math.ceil(self.num_bins / 2)
+            for param_name in self.grad_dists:
+                new_grad_dists = torch.zeros((self.num_chains, self.num_draws, new_bin_count), dtype=torch.int64).to(self.device)
+                for i in range(self.num_bins):
+                    new_grad_dists[:, :, i // 2] += self.grad_dists[param_name][:, :, i]
+                self.grad_dists[param_name] = new_grad_dists
+            self.num_bins = new_bin_count
+            self.bin_size *= 2
+
+    def sample(self):
+        return {
+            "gradient/distributions": self.grad_dists,
+        }
+
+    def plot(self, param_name: str, color='blue', plot_zero=True, chain: int = None):
+        grad_dist = self.grad_dists[param_name]
+        if chain is not None:
+            max_count = grad_dist[chain].max()
+        else:
+            max_count = grad_dist.sum(axis=0).max()
+
+        def get_color_alpha(count):
+            if count == 0:
+                return torch.tensor(0).to(self.device)
+            min_alpha = 0.05
+            max_alpha = 0.6
+            return (count / max_count) * (max_alpha - min_alpha) + min_alpha
+        
+        def build_rect(count, bin_min, bin_max, draw):
+            alpha = get_color_alpha(count)
+            pos = (draw, bin_min)
+            height = bin_max - bin_min
+            width = 1
+            return plt.Rectangle(pos, width, height, color=color, alpha=alpha.cpu().numpy().item(), linewidth=0)
+        
+        _, ax = plt.subplots()
+        for draw in range(self.num_draws):
+            for pos in range(self.num_bins):
+                bin_min = self.min_grad + pos * self.bin_size
+                bin_max = bin_min + self.bin_size
+                if chain is None:
+                    count = grad_dist[:, draw, pos].sum()
+                else:
+                    count = grad_dist[chain, draw, pos]
+                rect = build_rect(count, bin_min, bin_max, draw)
+                ax.add_patch(rect)
```

Comment:
Thanks for pointing this out. I did not try out imshow(), but I did change this to skip over rects with `count=0` and also to use `matplotlib.collections.PatchCollection` instead of adding patches one at a time. Consecutively, these two changes reduced runtime from around 100s -> 10s -> 2-3s.